### PR TITLE
[IP-299] Blockchain - Implement oracle contract retrieving ETH/USD rate

### PR DESCRIPTION
**Pull request by vladichhh**

### CODE REVIEW

File: contracts/Oracle/ExchangeOracle.sol [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+import "./../../node_modules/openzeppelin-solidity/contracts/ownership/Ownable.sol";
+import "./../../node_modules/openzeppelin-solidity/contracts/math/SafeMath.sol";
+
+
+contract ExchangeOracle is Ownable {
+
+    using SafeMath for uint256;
+
+    bool public isVaultitudeOracle = true;
```
Comment (Holymir) [inline contracts/Oracle/ExchangeOracle.sol]: unused variable?

File: contracts/Oracle/ExchangeOracle.sol [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+    uint256 public rate = 0;
```
Comment (Holymir) [inline contracts/Oracle/ExchangeOracle.sol]: you don't need to set default value

File: contracts/Oracle/ExchangeOracle.sol [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+     * By default rateMultiplier = 1000.
+     * With rate multiplier we can set the rate to be a float number.
+     *
+     * We use it as a multiplier because we can not pass float numbers in Ethereum.
+     * If the USD price becomes bigger than ether one, for example -> 1 USD = 10 ethers.
+     * We will pass 100 as rate and this will be relevant to 0.1 USD = 1 ether.
+     */
+    function setRateMultiplier(uint256 newRateMultiplier) external onlyOwner returns (bool) {
+        require(newRateMultiplier > 0);
+        require(newRateMultiplier % 10 == 0);
```
Comment (Holymir) [inline contracts/Oracle/ExchangeOracle.sol]: not sure if necessary. What if we set newRate to 5740?

File: contracts/Oracle/ExchangeOracle.sol [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+     *                  3999    1000
+     *      weiAmount = ---- * ------ * 1000000000000000000
+     *                  100    298758
+     *
+     *      return -> 133854156206695720 (in wei)
+     */
+    function calcWeiForUSDAmount(uint256 usdAmount) external view returns (uint256) {
+        require(usdAmount > 0);
+
+        uint256 weiAmount = usdAmount.mul(rateMultiplier);
```
Comment (Holymir) [inline contracts/Oracle/ExchangeOracle.sol]: maybe we should add to comments - usdAmount is passed multiplied by 100

File: contracts/Oracle/ExchangeOracle.sol [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+     */
+    function setRate(uint256 newRate) external onlyOwner returns (bool) {
+        require(newRate > 0);
+
+        uint256 oldRate = rate;
+        rate = newRate;
+
+        emit RateChanged(oldRate, newRate);
+
+        return true;
```
Comment (Holymir) [inline contracts/Oracle/ExchangeOracle.sol]: do we need to return bool?

File: contracts/Oracle/ExchangeOracle.sol [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        require(usdAmount > 0);
+
+        uint256 weiAmount = usdAmount.mul(rateMultiplier);
+        weiAmount = weiAmount.mul(ONE_ETHER_IN_WEI);
+        weiAmount = weiAmount.div(usdMultiplier);
+        weiAmount = weiAmount.div(rate);
+
+        return weiAmount;
+    }
+
```
[the remark this answers is not in the document]
Comment (vladichhh) [reply contracts/Oracle/ExchangeOracle.sol]: added function 'convertWeiToUSD'

File: contracts/Oracle/ExchangeOracle.sol [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+    /*
+     * Converts the specified wei amount in USD. The returned value is multiplied
+     * by the usdMultiplier value, which is by default 100.
+     *
+     *                  weiAmount             rate
+     *    usdAmount = ---------------- * -------------- * usdMultiplier
+     *                ONE_ETHER_IN_WEI   rateMultiplier
+     *
+     */
+    function convertWeiToUSD(uint256 weiAmount) external view returns (uint256) {
```
Comment (thcrnk) [inline contracts/Oracle/ExchangeOracle.sol]: Do we need a covertWeiToUSDByLastRate?

Comment (vladichhh) [reply contracts/Oracle/ExchangeOracle.sol]: added 'convertWeiToUSDByLastRate'

**Pull request merged**